Command line flag for default/custom configs.

bstein-clever opened the pull request:
I'm not the happiest with the error checking I added, or the related unit tests, but I'm not clear on a better way to handle it. Thoughts are welcome.

rgarcia commented on aviator/aviator.go:
```diff
@@ -125,15 +125,16 @@ func (a *Aviator) loadAviatorYAML(reader io.Reader, path string) error {
 
-func (a *Aviator) Plan(service string) error {
-	err := a.resolveEnvironment(service)
+// service = service name. useConfig = default, prod, etc
```
nit: comment should lint, i.e. be of the form "<function name> ..."

rgarcia commented on cmd/aviator/main.go:
```diff
@@ -22,6 +22,8 @@ func main() {
 	quiet := flag.Bool("quiet", false, "Sets log level to quiet")
 	list := flag.Bool("list", false, "List services that can be started")
+	useConfig := flag.String("config", "", "Start all services with a particular config (e.g. no prompts)")
+	defaults := flag.Bool("defaults", false, "Start all services with the default config (cannot be used with -config)")
```
I think it'd be clearer if we used the term "setup" instead of "config" (and renamed variables accordingly) to more closely match how you specify it in YAML

nathanleiby replied:
:+1:

bstein-clever replied:
I'll make it so.

rgarcia commented:
lgtm after a preliminary read, but I think @dfuentes or @nathanleiby should give the final word

side thought: the ability to specify the name of the setup you'd like to run makes me think the `default` designation in a setup's config might be kill-able as a feature. I.e. just have simple conventions around setup names e.g. `default`, `docker`, `staging`, etc. plus the ability to specify which setup you'd like to run.

nathanleiby commented on aviator/aviator.go:
```diff
@@ -229,22 +230,41 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
+	}
+	if isRealConfig == false {
+		log.Normalf("Unknown config: %s for service: %s", serv.selectedSetup, serv.Name)
+		// TODO(bstein): Throw an error here?
```
I would throw an error here, to be strict and clear.

Originally when we were making Aviator, we mainly thought of configs as corresponding to exactly one value. e.g. `staging` might mean `http://staging-api.example.com`. So we also thought it would be useful to be able to input something custom at runtime, i.e. literally the user would type `http://my-api.example.com` or `localhost:5000`. I don't believe this has been a commonly used or valuable feature.

nathanleiby replied:
Or rather than error, you could just prompt the user in a loop until they enter a valid value

bstein-clever replied:
No, I very much dislike the idea of an infinite loop - especially if providing a command-line default flag.
An infinite loop there would be bad, throwing an error is fine.

My worry about throwing an error is that it might potentially breaks current usage patterns, but perhaps not if things would fail anyway. I'll have to go read up on exceptions in Go.

nathanleiby replied:
Ah, yes, I was thinking more for when the user *isnt* using defaults. In that case, they shouldn't have to enter any setup names at all.

I was thinking a prompt loop might help if they user is on *non* default mode. If they type `stagng` instead of `staging`, and Aviator fails immediately, it seems painful. Especially if user has to enter multiple setups and makes a typo in the 3rd or 4th one. More preferable (to me) to re-prompt for a valid setup name.

nathanleiby commented on aviator/aviator.go:
```diff
@@ -229,22 +230,41 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
+		if name == serv.selectedSetup {
+			isRealConfig = true
+		}
+	}
```
Might be nice to use a `set` package. e.g. https://github.com/fatih/set

```go
validConfigs := set.New(serv.Setups...)
if ! validConfigs.Has(serv.selectedSetup) {
  // re-prompt user
}
```

bstein-clever replied:
Yes, much cleaner. I'm not too familiar with Go and libraries yet, so I'm stuck with basics.
Any good references/tutorials you'd suggest?

nathanleiby commented on aviator/aviator.go:
```diff
@@ -125,15 +125,16 @@ func (a *Aviator) loadAviatorYAML(reader io.Reader, path string) error {
 // service = service name. useConfig = default, prod, etc
-func (a *Aviator) Launch(service string) error {
-	if err := a.resolveEnvironment(service); err != nil {
+func (a *Aviator) Launch(service string, useConfig string) error {
```
nit: I would change the naming slightly. `useConfig` seems like a bool to me.

how about: `configName`

bstein-clever replied:
As it happens, the comment about that string predates these changes  - it was some old code of yours (f06fcacc), so I reused the name, since it seemed appropriate. "configName" to me sounds like something that's required.

If we decide to get rid of the default stuff, then we could use something like "defaultConfig"

rgarcia replied:
what about "setupName"?

nathanleiby replied:
:+1: for `setupName`

bstein-clever replied:
This ia a moot point if we're removing the option for a name choice entirely.

nathanleiby commented on cmd/aviator/main.go:
```diff
@@ -22,6 +22,8 @@ func main() {
 	trace := flag.Bool("trace", false, "Sets log level to trace")
 	quiet := flag.Bool("quiet", false, "Sets log level to quiet")
 	list := flag.Bool("list", false, "List services that can be started")
+	useConfig := flag.String("config", "", "Start all services with a particular config (e.g. no prompts)")
```
I'm cautious to have Aviator expect everyone to name their `setups` equivalently. 

I'm also thinking that a mixture of setup types (e.g. run one or two things locally, but point to everything else remotely on AWS) might be useful for debugging.

Do we have examples of people wanting this use case?

bstein-clever replied:
Well, there's an interesting debate here that should possibly be had in person.
Raf suggested getting rid of "default" entirely (it may be possible to have multiple defaults), and moving to consistent naming, while you suggest having split types could be useful.

My suggestion might be to allow having multiple tags for the same config, e.g. 
  default,bash:
        ...
  docker:
        ...
Or perhaps 
  default:
  bash:
        ...
  docker:
        ...

though that doesn't offer the choice for split configs. But if you're doing that, do you really want to use -setup at all? Wouldn't you specify command-line environment variables, or pick and choose manually, instead?

rgarcia replied:
I think we've opened a can of worms here...there is a number of directions we could go with this--I also think in the future you may want to pass into aviator the setups you'd like to use for every service, e.g. I want to run the service I'm working on from source, but I want all dependent services to run via docker.

I think we should leave this flag out for now and see how the -defaults flag works for people. Maybe we discover there are other setups that people want to run w/o interactive prompts.

bstein-clever replied:
Personally, I think those are distinct use cases, and passing in custom setups everywhere would be better as an entirely separate  flag.

I'll also observe that it's possible to set many of the parameters manually in the environment first - just the other day, Jian suggested the following to start the dashboard:
`MANDRILL_KEY=x CLEVER_RECAPTCHA_PUBLIC_KEY=x MIXPANEL_TOKEN=x CLEVER_RECAPTCHA_PRIVATE_KEY=x NODE_DEBUG=clever API_PATH=worker-clever-api-jian-1.ops.clever.com INSTANT_LOGIN_MONGO_URL=mongodb://mongodb-il-jian-dev.ops.clever.com/clever aviator clever-dashboard`

Which ends up only prompting for redis, as opposed to everything else, as well. But (I think, if I understand things correctly), that will only work if you're pointing everything to already running services, rather than having a split setup with some things running locally under different setups.

I think this is the point where I have to defer to someone else for the final decision, as I have the least experience with the tooling here.

rgarcia replied:
For now please remove this flag to keep this change confined to the original goal of adding the -defaults flag.

nathanleiby replied:
Agreed on simpler solution for now. Just `defaults`

nathanleiby commented on aviator/aviator.go:
```diff
@@ -229,22 +230,41 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
+	} else if useConfig != "default" {
+		// "default" falls through to the next case
+		// n.b. This means that if there's a config called "default", that is not itself the default,
+		// it won't get picked. But if someone does that, they diserve a debugging headache.
```
nit: s/diserve/deserve

bstein-clever replied:
Eek, thanks. Fixed.

nathanleiby commented on aviator/aviator_test.go:
```diff
@@ -175,11 +176,99 @@ func Test_DependencyCrawlingHarder(t *testing.T) {
+	input.WriteString("this_should_fail")
+	service.selectedSetup = ""
+	config = a.promptOrUseConfigForSetup(&service, "", &input)
+	assert.Equal(t, config.Command, "", "this should fail")
```
"this should fail" --> clarify why

bstein-clever replied:
Sorry, yes. This was meant as a placeholder until we'd decided on how to handle an invalid config, name and if we even wanted to fail fast there.

nathanleiby commented on aviator/aviator.go:
```diff
@@ -229,22 +230,41 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
-// promptForSetup prompts a user to choose which setup to use for a service.
-func (a *Aviator) promptForSetup(serv *Service) ServiceSetup {
+// promptOrUseConfigForSetup prompts a user to choose which setup to use for a service, or uses a config string, if provided.
+func (a *Aviator) promptOrUseConfigForSetup(serv *Service, useConfig string, input io.Reader) ServiceSetup {
```
I might separate into two methods:

1. the prompting logic (show this UI, keep repeating prompt until a valid value is given)
2. the default/useConfig logic (just choose a value immediately or error if no such setup exists)

That should make the unit testing easier, puts default logic in one place (should "default" be a setup name, or just a property?), and might lessen passing around the `useConfig` variable.

bstein-clever replied:
Sure, I can split it apart - that does make the promptOrUseConfigForSetup prototype cleaner, as well as the unit tests,
I considered it, but my worry was that I'd wanted to ensure from the tests that the "prompt..." function actually prompted somewhere.

bstein-clever replied:
Update: This no longer makes as much sense to split out if default is only a true/false, so I've not done that.

bstein-clever commented:
Ok, I've updated these as discussed - only the -default argument, and I've fixed the code to throw an error and have corrected tests.

Take another look and let me know if you have any further feedback. I'l squash these together once it's done, but for now the tree is slightly messy...

rgarcia commented on cmd/aviator/main.go:
```diff
@@ -107,9 +108,9 @@ func main() {
 
 	if *plan {
-		err = av.Plan(service)
+		err = av.Plan(service, &defaults)
```
I think this (and below) should be `*defaults`?

rgarcia commented on aviator/aviator.go:
```diff
@@ -229,23 +234,36 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
 
-// promptForSetup prompts a user to choose which setup to use for a service.
-func (a *Aviator) promptForSetup(serv *Service) ServiceSetup {
+// promptOrUseDefaultsForSetup prompts a user to choose which setup to use for a service, or uses a config string, if provided.
```
"... or uses a config string" comment out of date

rgarcia commented on aviator/aviator.go:
```diff
@@ -229,23 +234,36 @@ func (a *Aviator) resolveEnvironment(serviceName string) error {
+	// Ensure the chosen config name is valid.
+	_, validConfig := serv.Setups[serv.selectedSetup]
+	if !validConfig {
+		err := fmt.Errorf("Unknown config: %s for service: %s", serv.selectedSetup, serv.Name)
```
For consistency I'd use use "setup" instead of "config" in this error message and also `validSetup` instead of `validConfig`

rgarcia commented:
lgtm besides a few comments

rgarcia commented on aviator/aviator.go:
```diff
@@ -124,16 +124,18 @@ func (a *Aviator) loadAviatorYAML(reader io.Reader, path string) error {
-// service = service name. useConfig = default, prod, etc
-func (a *Aviator) Launch(service string) error {
-	if err := a.resolveEnvironment(service); err != nil {
+// Launch takes a service name. and whether to use the default setup.
```
I think you accidentally-d a sentence here :)

rgarcia commented on aviator/aviator.go:
```diff
@@ -124,16 +124,18 @@ func (a *Aviator) loadAviatorYAML(reader io.Reader, path string) error {
-func (a *Aviator) Plan(service string) error {
-	err := a.resolveEnvironment(service)
+// Plan takes a service name. and whether to use the default setup.
+// execute to run it (i.e. a dry run)
```
second line of this comment can be removed I think

also sentence weirdness in first line

bstein-clever replied:
Whoops, I think that got messed up in a merge conflict. Those should read:
// Launch takes a service name and whether to use the default setup, and executes it. 
// Plan takes a service name and whether to use the default setup, but does not execute it (i.e. a dry run)

rgarcia commented:
@bstein-clever assigning to you for edits. Please assign it back to me when ready!

bstein-clever commented:
Updated.

dfuentes commented:
:+1:

rgarcia commented:
lgtm!
